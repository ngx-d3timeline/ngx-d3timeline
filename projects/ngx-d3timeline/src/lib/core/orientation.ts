export enum Orientation {
  Horizontal = 'Horizontal',
  Vertical = 'Vertical'
}

export function flipOrientation(orientation: Orientation): Orientation {
  return orientation === Orientation.Vertical
    ? Orientation.Horizontal
    : Orientation.Vertical;
}
