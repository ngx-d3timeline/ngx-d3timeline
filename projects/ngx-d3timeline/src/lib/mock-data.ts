import { TimelineEvent } from './timeline-event';

export const mockData: TimelineEvent[] = [
  {
    id: 1,
    start: new Date('2017-09-23T08:45:00'),
    finish: new Date('2017-09-23T11:47:38'),
    type: 'Driving',
    series: 'Series 1'
  },
  {
    id: 2,
    start: new Date('2017-09-22T04:09:21'),
    finish: new Date('2017-09-22T08:39:21'),
    type: 'Driving',
    series: 'Series 1'
  },
  {
    id: 3,
    start: new Date('2017-09-22T08:39:21'),
    finish: new Date('2017-09-22T09:24:21'),
    type: 'DriveBreak',
    series: 'Series 1',
    sequence: 1
  },
  {
    id: 13,
    start: new Date('2017-09-22T10:00:00'),
    finish: new Date('2017-09-22T10:08:31'),
    type: 'Driving',
    series: 'Series 2'
  },
  {
    id: 14,
    start: new Date('2017-09-22T09:15:00'),
    finish: new Date('2017-09-22T09:16:16'),
    type: 'Waiting',
    series: 'Series 2',
    sequence: 6
  },
  {
    id: 15,
    start: new Date('2017-09-22T13:45:00'),
    finish: new Date('2017-09-22T14:52:27'),
    type: 'Driving',
    series: 'Series 2'
  },
  {
    id: 41,
    start: new Date('2017-09-23T08:45:00'),
    finish: new Date('2017-09-23T11:47:38'),
    type: 'Driving',
    series: 'Series 3'
  },
  {
    id: 42,
    start: new Date('2017-09-22T04:09:21'),
    finish: new Date('2017-09-22T08:39:21'),
    type: 'Driving',
    series: 'Series 3'
  },
  {
    id: 43,
    start: new Date('2017-09-22T08:39:21'),
    finish: new Date('2017-09-22T09:24:21'),
    type: 'DriveBreak',
    series: 'Series 3',
    sequence: 1
  }
];
