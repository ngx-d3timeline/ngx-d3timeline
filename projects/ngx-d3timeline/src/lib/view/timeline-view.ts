export class TimelineView {
  readonly width: number;
  readonly height: number;
  readonly margin: number;

  readonly bottom: number;
  readonly left: number;
  readonly right: number;
  readonly top: number;

  constructor({
    width,
    height,
    margin
  }: {
    width: number;
    height: number;
    margin: number;
  }) {
    this.width = width;
    this.height = height;
    this.margin = margin;

    this.bottom = height - margin;
    this.left = 0;
    this.right = width - margin;
    this.top = 0;
  }
}
