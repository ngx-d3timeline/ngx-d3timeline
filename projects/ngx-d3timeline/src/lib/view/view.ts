export class View {
  readonly width: number;
  readonly height: number;
  readonly margin = 50;

  readonly left = this.margin;
  readonly top = this.margin;
  readonly right: number;
  readonly bottom: number;

  constructor([width, height]: [number, number]) {
    this.width = width;
    this.height = height;

    this.bottom = height - this.margin;
    this.right = width - this.margin;
  }
}
