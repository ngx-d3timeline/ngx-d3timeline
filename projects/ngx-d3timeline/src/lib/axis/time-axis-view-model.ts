import { scaleTime, ScaleTime } from 'd3-scale';
import { max, min } from 'd3-array';
import { TimelineEvent } from '../timeline-event';
import { TimelineView } from '../view/timeline-view';

export class TimeAxisViewModel {
  readonly scaleTime: ScaleTime<number, number>;

  constructor(data: TimelineEvent[], timelineView: TimelineView, event: any) {
    this.scaleTime = this.configureScaleTime(data, timelineView, event);
  }

  get tickInfos(): { label: string; transform: string }[] {
    return this.scaleTime.ticks().map(tick => ({
      label: this.scaleTime.tickFormat()(tick),
      transform: this.tickTransform(tick)
    }));
  }

  get rangeLimit() {
    return this.scaleTime.range()[1];
  }

  private configureScaleTime(
    data: TimelineEvent[],
    timelineView: TimelineView,
    event: any
  ) {
    const scale = scaleTime()
      .domain([min(data, d => d.start), max(data, d => d.finish)])
      .range([0, timelineView.bounds.bottom]);

    return event ? event.transform.rescaleY(scale) : scale;
  }

  private tickTransform(tick: Date) {
    return `translate(0,${this.scaleTime(tick)})`;
  }
}
