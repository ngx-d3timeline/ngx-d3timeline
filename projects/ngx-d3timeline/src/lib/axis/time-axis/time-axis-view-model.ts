import { scaleTime, ScaleTime } from 'd3-scale';
import { max, min } from 'd3-array';
import { TimelineEvent } from '../../timeline-event';
import { TimelineView } from '../../view/timeline-view';
import { Orientation } from '../../orientation';

export class TimeAxisViewModel {
  readonly scaleTime: ScaleTime<number, number>;

  constructor(
    data: TimelineEvent[],
    timelineView: TimelineView,
    event: any,
    private readonly orientation: Orientation
  ) {
    this.scaleTime = this.configureScaleTime(data, timelineView, event);
  }

  get tickInfos(): { label: string; transform: string }[] {
    return this.scaleTime.ticks().map(tick => ({
      label: this.scaleTime.tickFormat()(tick),
      transform: this.tickTransform(tick)
    }));
  }

  get rangeLimit() {
    return this.scaleTime.range()[1];
  }

  private configureScaleTime(
    data: TimelineEvent[],
    timelineView: TimelineView,
    event: any
  ) {
    const scale = scaleTime()
      .domain([min(data, d => d.start), max(data, d => d.finish)])
      .range(this.getRange(timelineView));

    return event ? this.getTransformedScale(scale, event) : scale;
  }

  private getRange(view: TimelineView) {
    return this.orientation === Orientation.Vertical
      ? [view.top, view.bottom]
      : [view.left, view.right];
  }

  private tickTransform(tick: Date) {
    return this.orientation === Orientation.Vertical
      ? `translate(0,${this.mapToPixel(tick)})`
      : `translate(${this.mapToPixel(tick)}, 0)`;
  }
  private getTransformedScale(scale: ScaleTime<number, number>, event: any) {
    return this.orientation === Orientation.Vertical
      ? event.transform.rescaleY(scale)
      : event.transform.rescaleX(scale);
  }

  mapToPixel(domain: Date): number {
    return this.scaleTime(domain);
  }
}
