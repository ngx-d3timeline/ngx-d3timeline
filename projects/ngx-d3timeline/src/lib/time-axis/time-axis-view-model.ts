import { scaleTime, ScaleTime } from 'd3-scale';
import { max, min } from 'd3-array';
import { TimelineEvent } from '../timeline-event';
import { TimelineView } from '../view/timeline-view';
import { Orientation } from '../orientation';

export class TimeAxisViewModel {
  readonly scaleTime: ScaleTime<number, number>;

  constructor(
    data: TimelineEvent[],
    timelineView: TimelineView,
    event: any,
    private readonly orientation: Orientation
  ) {
    this.scaleTime = this.configureScaleTime(data, timelineView, event);
  }

  get tickInfos(): { label: string; transform: string }[] {
    return this.scaleTime.ticks().map(tick => ({
      label: this.scaleTime.tickFormat()(tick),
      transform: this.tickTransform(tick)
    }));
  }

  get rangeLimit() {
    return this.scaleTime.range()[1];
  }

  private configureScaleTime(
    data: TimelineEvent[],
    timelineView: TimelineView,
    event: any
  ) {
    const scale = scaleTime()
      .domain([min(data, d => d.start), max(data, d => d.finish)])
      .range(this.getRange(timelineView.bounds));

    return event ? this.getTransformedScale(scale, event) : scale;
  }

  private getRange(bounds: any) {
    return this.orientation === Orientation.Vertical
      ? [bounds.top, bounds.bottom]
      : [bounds.left, bounds.right];
  }

  private tickTransform(tick: Date) {
    return this.orientation === Orientation.Vertical
      ? `translate(0,${this.scaleTime(tick)})`
      : `translate(${this.scaleTime(tick)}, 0)`;
  }
  private getTransformedScale(scale: ScaleTime<number, number>, event: any) {
    return this.orientation === Orientation.Vertical
      ? event.transform.rescaleY(scale)
      : event.transform.rescaleX(scale);
  }
}
