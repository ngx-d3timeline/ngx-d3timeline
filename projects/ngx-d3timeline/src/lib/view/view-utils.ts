import { View } from './view';
import { Point } from '../core/point';
import { Rectangle } from '../core/rectangle';

export function getViewTopLeft(view: View): Point {
  return { x: view.left, y: view.top };
}

export function getViewClipRectangle(view: View): Rectangle {
  return {
    width: view.width - 2 * view.margin,
    height: view.height - 2 * view.margin,
    x: view.left,
    y: view.top
  };
}
