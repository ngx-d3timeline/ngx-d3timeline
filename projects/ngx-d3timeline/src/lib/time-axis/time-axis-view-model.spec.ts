import { TimeAxisViewModel } from './time-axis-view-model';
import { TimelineView } from '../view/timeline-view';
import { mockData } from '../mock-data';
import { Orientation } from '../orientation';

describe('TimeAxisViewModel', () => {
  let vm: TimeAxisViewModel;
  const timelineView: TimelineView = new TimelineView(300, 400, 50);

  describe('when Vertical', () => {
    beforeEach(() => {
      vm = new TimeAxisViewModel(
        mockData,
        timelineView,
        null,
        Orientation.Vertical
      );
    });

    it('should return correct tickInfo', () => {
      expect(vm.tickInfos).toMatchSnapshot();
    });

    it('should return correct rangeLimit', () => {
      expect(vm.rangeLimit).toBe(timelineView.bounds.bottom);
    });
  });

  describe('when Horizontal', () => {
    beforeEach(() => {
      vm = new TimeAxisViewModel(
        mockData,
        timelineView,
        null,
        Orientation.Horizontal
      );
    });

    it('should return correct tickInfo', () => {
      expect(vm.tickInfos).toMatchSnapshot();
    });

    it('should return correct rangeLimit', () => {
      expect(vm.rangeLimit).toBe(timelineView.bounds.right);
    });
  });
});
