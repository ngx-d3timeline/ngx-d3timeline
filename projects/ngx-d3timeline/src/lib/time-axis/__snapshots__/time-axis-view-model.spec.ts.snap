// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`TimeAxisViewModel when Horizontal should return correct tickInfo 1`] = `
Array [
  Object {
    "label": "06 AM",
    "transform": "translate(14.572376796579364, 0)",
  },
  Object {
    "label": "09 AM",
    "transform": "translate(38.27800556643283, 0)",
  },
  Object {
    "label": "12 PM",
    "transform": "translate(61.98363433628629, 0)",
  },
  Object {
    "label": "03 PM",
    "transform": "translate(85.68926310613975, 0)",
  },
  Object {
    "label": "06 PM",
    "transform": "translate(109.39489187599321, 0)",
  },
  Object {
    "label": "09 PM",
    "transform": "translate(133.10052064584667, 0)",
  },
  Object {
    "label": "Sat 23",
    "transform": "translate(156.80614941570013, 0)",
  },
  Object {
    "label": "03 AM",
    "transform": "translate(180.51177818555362, 0)",
  },
  Object {
    "label": "06 AM",
    "transform": "translate(204.21740695540709, 0)",
  },
  Object {
    "label": "09 AM",
    "transform": "translate(227.92303572526055, 0)",
  },
]
`;

exports[`TimeAxisViewModel when Vertical should return correct tickInfo 1`] = `
Array [
  Object {
    "label": "06 AM",
    "transform": "translate(0,20.40132751521111)",
  },
  Object {
    "label": "09 AM",
    "transform": "translate(0,53.58920779300596)",
  },
  Object {
    "label": "12 PM",
    "transform": "translate(0,86.77708807080081)",
  },
  Object {
    "label": "03 PM",
    "transform": "translate(0,119.96496834859565)",
  },
  Object {
    "label": "06 PM",
    "transform": "translate(0,153.1528486263905)",
  },
  Object {
    "label": "09 PM",
    "transform": "translate(0,186.34072890418534)",
  },
  Object {
    "label": "Sat 23",
    "transform": "translate(0,219.52860918198022)",
  },
  Object {
    "label": "03 AM",
    "transform": "translate(0,252.71648945977506)",
  },
  Object {
    "label": "06 AM",
    "transform": "translate(0,285.9043697375699)",
  },
  Object {
    "label": "09 AM",
    "transform": "translate(0,319.09225001536475)",
  },
]
`;
